### DISCUSSION TITLE
How best to loop through folder with files that contain list of AIDs for batch RTR processing

### DISCUSSION THREAD

**Discussion by b1tf0rbyt3s** [CrowdStrike/falconpy#1142] [Q&A]
So, what I am trying to do is loop through a number of text files that contain a comma separated list of AIDs and feed that into a variable to use for initiating batch RTR sessions via BatchInitSessions.  It works exactly as you would expect with the first run through but then I get a 400 invalid batch ID error when it loops back - "invalid batch ID provided - Status Code: 400" and the batch ID will be empty when printing it to stdout.  I get a batch ID when it runs through first time and successfully runs a series of put and runs. But then it will give me the 400 batch error as it loops back through with the next file and there will be no batch ID when it grabs the second file (The text files containing the list of AIDs are retrieved programmatically via API so they are identical).  

Not sure what I am doing wrong or how to go about troubleshooting why I get this error.  Or if there is a better way to do this?

Update: Loop works if I define the function within the same python file - as opposed to having the function in a separate python file and importing.  Just not sure why it works one way and not the other?

```python
# Function defined in falcon_put_and_run.py in modules folder
def execute_batch(falcon_session, target_hosts):
    """
    Initialize a batch of RTR Sessions.
    """
    init_body = {"host_ids": target_hosts, "queue_offline": False}

    session_init = falcon_session.command("BatchInitSessions", timeout=30, timeout_duration="s", body=init_body)
    batch_id = session_init["body"]["batch_id"]
    print(f"Batch ID: {batch_id}")
    #Additional code in function follows to do a series of put and runs via BatchAdminCmd

# main.py
from mods import falcon_put_and_run
from falconpy import APIHarnessV2

falcon = APIHarnessV2(client_id=environ.get('<REDACTED>'),
                          client_secret=environ.get('<REDACTED>'))

host_ids_path = 'working/host_ids/'
# Loop through folder and ingest the AIDs for Batch RTR session - each file will have up to 1k AIDs
for id_file_name in os.listdir(host_ids_path):
    with open(f"{host_ids_path}{id_file_name}", "r") as host_id_file:
        host_ids = host_id_file.read().strip().split(",")
    falcon_put_and_run.execute_batch(falcon, host_ids)


```

**Comment by jshcodes** [comment]
If we turn on debugging, do we get more detail when the second file is processed and the `execute_batch` function calls [__`BatchInitSessions`__](https://www.falconpy.io/Service-Collections/Real-Time-Response.html#batchinitsessions)?

```
# main.py
from mods import falcon_put_and_run
from falconpy import APIHarnessV2
import logging
logging.basicConfig(level=logging.DEBUG)

falcon = APIHarnessV2(client_id=environ.get('<REDACTED>'),
                      client_secret=environ.get('<REDACTED>'),
                      debug=True
                      )

# etc...
```
